src/home/room_image_viewer.rs

```
@@ -0,0 +1,92 @@
+        }
+        (MediaCacheEntry::Failed(status_code), MediaFormat::File) => {
+            let error = match status_code {
+                StatusCode::NOT_FOUND => ImageViewerError::NotFound,
+                StatusCode::INTERNAL_SERVER_ERROR => ImageViewerError::ConnectionFailed,
+                StatusCode::PARTIAL_CONTENT => ImageViewerError::BadData,
+                StatusCode::UNAUTHORIZED => ImageViewerError::Unauthorized,
+                StatusCode::REQUEST_TIMEOUT => ImageViewerError::Timeout,
+                _ => ImageViewerError::Unknown,
+            };
+            cx.action(ImageViewerAction::Show(LoadState::Error(error)));
+            // Remove failed media entry from cache for MediaFormat::File so as to start all over again from loading Thumbnail.
+            media_cache.remove_cache_entry(&mxc_uri, Some(MediaFormat::File));
+        }
+        _ => {}
+    }
+}
+
+/// Extracts image name and size from an event timeline item.
+pub fn extract_image_info(event_tl_item: &EventTimelineItem) -> (String, i32) {
+    if let Some(message) = event_tl_item.content().as_message() {
+        if let MessageType::Image(image_content) = message.msgtype() {
+            let name = message.body().to_string();
+            let size = image_content
+                .info
+                .as_ref()
+                .and_then(|info| info.size)
+                .map(|s| i32::try_from(s).unwrap_or_default())
+                .unwrap_or(0);
+            (name, size)
+        } else {
+            ("Unknown Image".to_string(), 0)
+        }
+    } else {
+        ("Unknown Image".to_string(), 0)
+    }
+}
+
+/// Condensed message does not have a profile, so we need to find the previous portal list item.
+/// Searches backwards for a non-empty display name and avatar in previous portal list items.
+/// Mutates display_name and avatar_ref.
+/// Returns when first non-empty display name is found.
+pub fn find_previous_profile_in_condensed_message(
```

1. Reading widget values from the UI is an odd way to try to obtain content for a timeline event. I'm not sure what context you have when calling this, but why not just get this info from the event timeline item itself?
2. Using this style of output parameters is strange and not very rusty. I'd recommend you just return the display_name and the AvatarRef directly instead of mutating them, because that would clearly convey when you were _unable_ to obtain them. Currently, with this function signature, there's no way to know if you successfully obtained those two bits of data, which is a form of silencing errors.
3. Nit: doc comments are rendered using markdown, so the line separation here gets lost. Generally, you should start the doc comment with a short one-line summary, then a blank line, and then the details in any format you wish.